### PR TITLE
Extract render policy decisions

### LINKED ISSUE #356: Implement #340: isolate render-budget decisions
Child issue of #340.

## Goal

Make the remaining render-policy decisions readable and independently testable by separating pure state decisions from the LuCI view’s DOM orchestration.

## Scope

- Extract the UI-independent render-policy decisions currently embedded in `view/status/fwlive.js`, specifically the weak-device display cap and the per-render cost calculation.
- Give the extracted helper an explicit input/output contract: row limit, weak-device state/cap, prior rendered-row identity/count, and last-batch new-ID count in; visible-row cap and render cost out.
- Keep token-bucket mutation, scheduling, DOM updates, table rendering, polling, buffer application, wire format, persistence, and user-facing behavior unchanged.
- Add module-level tests for unchanged boundaries and existing view tests for integration.
- Preserve the `tablePaused` / hidden / in-flight independent axes and the `lastBatchNewIdCount` meaning from #354.

## Non-goals

- No virtualization or weak-device threshold changes.
- No renderer/table refactor, classifier work, ACL changes, or #209 work.
- No new abstraction without a concrete pure decision contract.

## Acceptance

- The extracted module has one clear responsibility and no DOM dependency.
- The view delegates both decisions without duplicating their logic.
- Existing render-throttle and weak-device behavior tests remain green; module tests cover unchanged, visible-count-change, and batch-cost boundaries.
- Full repository validation passes.

Refs #340

### PR DESCRIPTION
**Pull request by lucas-albers-lz4**
## Summary

- Extract pure weak-device display-cap and render-cost decisions into `fwlive.render-policy`.
- Keep token-bucket mutation, scheduling, DOM updates, table rendering, polling, buffering, thresholds, and user-visible behavior unchanged.
- Add module-level boundary tests and register them in the full test suite.

## Validation

- `node tests/fwlive-render-policy.test.js`
- Existing module, view, lint, and syntax checks
- `./scripts/fwlive-test.sh`

Closes #356
Refs #340

## Summary by CodeRabbit

- **New Features**
  - Improved live-view rendering on weaker devices by limiting the number of displayed rows according to device capability and user settings.
  - Refined rendering-cost decisions to support more efficient updates when displayed data changes.

- **Documentation**
  - Updated architecture and package documentation to describe the rendering-policy component.

- **Tests**
  - Added coverage for row limits, update-cost calculations, and live-view integration.

### CODE REVIEW

File: tests/fwlive-render-policy.test.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+assert.strictEqual(
+	renderCost({
+		visibleRowCount: 1,
+		visibleHeadId: 'b',
+		lastRenderedRowCount: 1,
+		lastRenderedHeadId: 'a',
+		lastBatchNewIdCount: 0
+	}),
+	1,
+	'new head always has a minimum cost of one'
```
Comment (lucas-albers-lz4) [inline tests/fwlive-render-policy.test.js:128]: Low: count-change is never tested when the head also moves. That is the Limit/trim + live-events case the old comment existed for:

```js
assert.strictEqual(
	renderCost({
		visibleRowCount: 2,
		visibleHeadId: 'b',
		lastRenderedRowCount: 1,
		lastRenderedHeadId: 'a',
		lastBatchNewIdCount: 7
	}),
	1,
	'count change must beat batch cost when the head also moves'
);
```

Same gap on the identity path: unchanged count+head with `lastBatchNewIdCount: 7` must stay **0**, or a later reorder of the cheap-identity check vs batch cost will not fail these tests.

File: openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * The caller supplies view state and receives the visible-row cap plus the
+ * cost of the next render. This module does not mutate state, schedule work,
+ * inspect the DOM, or know how rows are rendered.
+ *
+ * @param {object} opts
+ * @param {number} opts.rowLimit - User-selected stored-row limit.
+ * @param {boolean} opts.weakDevice - Whether the weak-device cap is active.
+ * @param {number} opts.weakDeviceDisplayRowCap - Maximum visible rows when capped.
+ * @returns {number} Maximum rows to render.
+ */
```
Comment (lucas-albers-lz4) [inline openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js:16]: Nit: this JSDoc is still the combined `decide()` contract ("visible-row cap plus the cost of the next render" plus `@param`s that belong only to `displayRowCap`). After the split, keep a short module comment here and let `displayRowCap` own its own params/returns.

File: openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+	/* This cheap identity check intentionally does not compare full row content. */
+	if (!count && !lastRenderedRowCount) return 0;
+	if (count === lastRenderedRowCount && headId === lastRenderedHeadId) return 0;
+
+	/* Count changes must stay cheap so Limit/filter/trim updates remain paintable. */
+	if (count !== lastRenderedRowCount) renderCost = 1;
+	else renderCost = Math.max(1, opts.lastBatchNewIdCount || 1);
+
+	return renderCost;
```
Comment (lucas-albers-lz4) [inline openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js]: Nit: `function renderCost` plus `let renderCost` is leftover from the if/else chain. The early returns already exist, so this can be:

```js
if (count !== lastRenderedRowCount) return 1;
return Math.max(1, opts.lastBatchNewIdCount || 1);
```

**Review by lucas-albers-lz4 — COMMENTED** [review]
No blocking issues. This is a clean extract: the view still owns row identity and token-bucket mutation, `fwlive.render-policy` only answers cap and cost, and CI is green.

The second commit is the right shape. A combined `decide()` forced dummy cost inputs into `displayRowCap()`; splitting the two calls matches #356.

**Summary of inline notes**

- Low: pin count-change vs batch-cost when the head also moves, and identity vs a nonzero `lastBatchNewIdCount`.
- Nit: leftover combined-`decide()` JSDoc and `function renderCost` / `let renderCost` shadowing.

**Not on the diff (so not inline)**

`docs/developer/architecture.md` (module-split table) and `openwrt-feed/luci-app-fwlive/README.md` (package layout) were updated for `poll-coordinator.js` and still omit `render-policy.js`. `architecture.md` owns that list.

I would merge after the count-vs-head test (and the doc/JSDoc nits if you want the extract to match the poll-coordinator PR).

**Comment by lucas-albers-lz4** [comment]
# Multi-Model Review — lucas-albers-lz4/fwlive PR #357

> **Head SHA:** `edc77a26f418ad06d12c154a466efdb6d662d232`  •  **Reviewers:** muse-spark-1.3-contributor, glm-5.3-flash, deepseek-v4.1-flash  •  **Synthesized by:** deepseek-v4-pro
> **Verdict:** APPROVE

## TL;DR
PR extracts two pure view decisions (`displayRowCap` and the render-cost throttle) from `view/status/fwlive.js` into a new `fwlive/render-policy.js` baseclass module, wires it into both Node test harnesses, and adds boundary tests registered in `scripts/fwlive-test.sh`. All three reviewers independently confirmed the extraction is behavior-preserving and that the touched tests pass locally. No critical or major findings; the consensus is a handful of minor/nit items (the `renderCost` variable shadowing is the only one flagged by all three, and it is readability-only). Safe to merge as-is; addressing the minor items would be polish.

## Findings (sorted by severity)

### 🔴 Critical
None.

### 🟡 Major
None.

### 🟢 Minor
- **Finding:** Local `let renderCost;` shadows the module-level `function renderCost` inside its own body.
- **File:** `openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js:42`
- **Reviewers who flagged:** A, B, C
- **Problem:** Legal but confusing; invites `no-shadow` lint failures and misreading. Harmless today (never recursive), a trap if the function ever references itself.
- **Suggested fix:** Rename the local to `cost` (A, B) or return each branch directly and drop the mutable binding (A, C).
- **Confidence:** high

- **Finding:** `displayRowCap` does no input validation — missing/non-numeric `rowLimit` yields `undefined`/`NaN` that propagates downstream.
- **File:** `openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js:19-22`
- **Reviewers who flagged:** A
- **Problem:** Pure-policy module silently produces an unusable cap instead of failing fast on a programmer error; the caller must remember to pass valid numbers.
- **Suggested fix:** Coerce with a `Number.isFinite` check/default, or explicitly document the caller-must-supply contract.
- **Confidence:** medium

### 💡 Nits
- **Finding:** `assert.deepStrictEqual` used to compare a primitive number.
- **File:** `tests/fwlive-render-policy.test.js:42`
- **Reviewers who flagged:** A, C
- **Suggested fix:** Use `assert.strictEqual(...)` to match surrounding assertions.
- **Confidence:** high

- **Finding:** Module JSDoc block mixes file-level prose with `@param`/`@returns` tags that actually describe `displayRowCap`.
- **File:** `openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js:6-18`
- **Reviewers who flagged:** B, C
- **Suggested fix:** Split into a file-header comment plus a `displayRowCap`-specific JSDoc (or trim the block to prose only).
- **Confidence:** medium/high

- **Finding:** Header comments run directly into `'require baseclass';` with no blank separator, unlike sibling modules (e.g. `poll-coordinator.js`).
- **File:** `openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js:3-4`
- **Reviewers who flagged:** B
- **Suggested fix:** Add a blank line before the require statement.
- **Confidence:** high

- **Finding:** README package-layout table does not list the new `render-policy.js` module (precedent #353 added a row for `poll-coordinator.js`).
- **File:** `openwrt-feed/luci-app-fwlive/README.md:19-21`
- **Reviewers who flagged:** C
- **Suggested fix:** Add a `render-policy.js` row to the module table.
- **Confidence:** high

- **Finding:** New test block in `scripts/fwlive-test.sh` lost its separating blank line from the modules-smoke block.
- **File:** `scripts/fwlive-test.sh:133-135`
- **Reviewers who flagged:** B
- **Suggested fix:** Restore the blank line before `echo "== fwlive render policy =="`.
- **Confidence:** high

- **Finding:** `|| 0` / `|| ''` normalization mishandles falsy-but-valid identities (e.g. numeric `0`) and negative counts.
- **File:** `openwrt-feed/luci-app-fwlive/htdocs/luci-static/resources/fwlive/render-policy.js:38-41`
- **Reviewers who flagged:** A
- **Suggested fix:** Use `??` / explicit `Number.isFinite` checks instead of `||` fallbacks.
- **Confidence:** low

- **Finding:** `render-policy` not added to the extracted-modules smoke section of `tests/fwlive-modules-smoke.test.js`.
- **File:** `tests/fwlive-modules-smoke.test.js:88-91`
- **Reviewers who flagged:** C
- **Suggested fix:** Add a one-line smoke assertion, or note the dedicated test supersedes it.
- **Confidence:** medium

## Reviewer agreement matrix

| Finding | A | B | C | Notes |
|---|---|---|---|---|
| `let renderCost` shadows function name | ✓ | ✓ | ✓ | All three flag the same line |
| `displayRowCap` lacks input validation | ✓ | — | — | A only |
| `deepStrictEqual` on a primitive | ✓ | — | ✓ | Same line, same fix |
| JSDoc block misattached | — | ✓ | ✓ | B/C both flag prose+param mix |
| Missing blank line before require | — | ✓ | — | B only |
| README module table missing row | — | — | ✓ | C only |
| test.sh blank-line drift | — | ✓ | — | B only |
| `|| 0`/`|| ''` falsy-id handling | ✓ | — | — | A nit, low confidence |
| modules-smoke omission | — | — | ✓ | C only |

## What the reviewers disagreed about
No substantive disagreements. All three agree the extraction is behavior-preserving and the shadowing is readability-only. Minor divergence is only in fix style: A and C prefer "return directly per branch" while B prefers "rename to `cost`" for the shadowing issue — a stylistic tie, human call (renaming is the smaller diff).

## Things the reviewers may have missed
- None beyond the above. The diff is small and self-contained; the only untested surface (real LuCI module loader / Playwright `test:view` fixture) was explicitly called out by B and C as "couldn't verify" rather than as an actual defect.

## Suggested next actions for the author
- [ ] Rename the `let renderCost` local to `cost` (or return directly) to clear the triple-flagged shadow.
- [ ] Change `assert.deepStrictEqual` → `assert.strictEqual` at `tests/fwlive-render-policy.test.js:42`.
- [ ] Add the `render-policy.js` row to the README module table (matches #353 convention).
- [ ] Split the module-level JSDoc from the `displayRowCap` JSDoc (and add the blank line before `'require baseclass';`).
- [ ] (Optional) Add a `Number.isFinite` guard to `displayRowCap`, or document the caller contract.

## Reviewer confidence summary
- Reviewer A (muse-spark-1.3-contributor): Thorough on edge cases and test coverage; ran only the new test + `node --check` (eslint deps missing), hedged on full suite.
- Reviewer B (glm-5.3-flash): Ran new test + module smoke tests; verified both load-path harnesses in lockstep; hedged on real LuCI loader and renderer/xss harness registration.
- Reviewer C (deepseek-v4.1-flash): Most comprehensive validation (lint:js, lint:format, new test, modules-smoke, four `loadFwliveView`-backed tests); hedged on Playwright browser and full `fwlive-test.sh` (needs ruff/shellcheck/jshn).

## Process notes
- Wall time: n/a (offline synthesis)
- Reviewer failures: none (all three produced reports)
- Diff was truncated: no (320-line diff read in full)
- Total cost: low — small, single-concern PR; all reviewers converged with no critical/major findings.
